pkg/ottl/contexts/internal/map_test.go

```diff
@@ -0,0 +1,236 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/ottltest"
+)
+
+func Test_GetMapValue_Invalid(t *testing.T) {
+	tests := []struct {
+		name string
+		keys []ottl.Key
+		err  error
+	}{
+		{
+			name: "no keys",
+			keys: []ottl.Key{},
+			err:  fmt.Errorf("cannot get map value without key"),
+		},
+		{
+			name: "first key not a string",
+			keys: []ottl.Key{
+				{
+					Int: ottltest.Intp(0),
+				},
+			},
+			err: fmt.Errorf("non-string indexing is not supported"),
+		},
+		{
+			name: "index map with int",
+			keys: []ottl.Key{
+				{
+					String: ottltest.Strp("map"),
+				},
+				{
+					Int: ottltest.Intp(0),
+				},
+			},
+			err: fmt.Errorf("map must be indexed by a string"),
+		},
+		{
+			name: "index slice with string",
+			keys: []ottl.Key{
+				{
+					String: ottltest.Strp("slice"),
+				},
+				{
+					String: ottltest.Strp("invalid"),
+				},
+			},
+			err: fmt.Errorf("slice must be indexed by an int"),
+		},
+		{
+			name: "index too large",
+			keys: []ottl.Key{
+				{
+					String: ottltest.Strp("slice"),
+				},
+				{
+					Int: ottltest.Intp(1),
+				},
+			},
+			err: fmt.Errorf("index 1 out of bounds"),
+		},
+		{
+			name: "index too small",
+			keys: []ottl.Key{
+				{
+					String: ottltest.Strp("slice"),
+				},
+				{
+					Int: ottltest.Intp(-1),
+				},
+			},
+			err: fmt.Errorf("index -1 out of bounds"),
+		},
+		{
+			name: "invalid type",
+			keys: []ottl.Key{
+				{
+					String: ottltest.Strp("string"),
+				},
+				{
+					String: ottltest.Strp("string"),
+				},
+			},
+			err: fmt.Errorf("type Str does not support string indexing"),
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			m := pcommon.NewMap()
+			m.PutStr("string", "invalid")
+			m.PutEmptyMap("map").PutStr("foo", "bar")
+
+			s := m.PutEmptySlice("slice")
+			s.AppendEmpty()
+
+			_, err := GetMapValue(m, tt.keys)
+			assert.Equal(t, tt.err, err)
+		})
+	}
+}
+
+func Test_GetMapValue_MissingKey(t *testing.T) {
```

Comment:
Two similar "missing key" cases I think we should cover for `SetMapValue`:
1. Setting nested keys in an empty map
2. Setting nested keys inside a map with a nested empty map.